tsdb/engine/tsm1/engine.go

```diff
@@ -127,58 +127,81 @@ func (e *DevEngine) Close() error {
 func (e *DevEngine) SetLogOutput(w io.Writer) {}
 
 // LoadMetadataIndex loads the shard metadata into memory.
-func (e *DevEngine) LoadMetadataIndex(shard *tsdb.Shard, index *tsdb.DatabaseIndex, measurementFields map[string]*tsdb.MeasurementFields) error {
+func (e *DevEngine) LoadMetadataIndex(_ *tsdb.Shard, index *tsdb.DatabaseIndex, measurementFields map[string]*tsdb.MeasurementFields) error {
 	keys := e.FileStore.Keys()
-	for _, k := range keys {
-		seriesKey, field := seriesAndFieldFromCompositeKey(k)
-		measurement := tsdb.MeasurementFromSeriesKey(seriesKey)
 
-		m := index.CreateMeasurementIndexIfNotExists(measurement)
-		m.SetFieldName(field)
+	keysLoaded := make(map[string]bool)
 
+	for _, k := range keys {
 		typ, err := e.FileStore.Type(k)
 		if err != nil {
 			return err
 		}
+		fieldType, err := tsmFieldTypeToInfluxQLDataType(typ)
+		if err != nil {
+			return err
+		}
 
-		mf := measurementFields[measurement]
-		if mf == nil {
-			mf = &tsdb.MeasurementFields{
-				Fields: map[string]*tsdb.Field{},
-			}
-			measurementFields[measurement] = mf
+		if err := e.addToIndexFromKey(k, fieldType, index, measurementFields); err != nil {
+			return err
 		}
 
-		switch typ {
-		case BlockFloat64:
-			if err := mf.CreateFieldIfNotExists(field, influxql.Float, false); err != nil {
-				return err
-			}
-		case BlockInt64:
-			if err := mf.CreateFieldIfNotExists(field, influxql.Integer, false); err != nil {
-				return err
-			}
-		case BlockBool:
-			if err := mf.CreateFieldIfNotExists(field, influxql.Boolean, false); err != nil {
-				return err
-			}
-		case BlockString:
-			if err := mf.CreateFieldIfNotExists(field, influxql.String, false); err != nil {
-				return err
-			}
-		default:
-			return fmt.Errorf("unkown block type for: %v. got %v", k, typ)
+		keysLoaded[k] = true
+	}
+
+	// load metadata from the Cache
+	e.Cache.mu.Lock() // shouldn't need the lock, but just to be safe
+	defer e.Cache.mu.Unlock()
+
+	for key, entry := range e.Cache.store {
```

Comment:
I can update that, but it's tricky. Either I have to make a complete copy, or I make the `Cache` implement Locker as you mentioned. But then the store is accessible with no guarantees that it's safe.